cmd/restic/cmd_copy.go

```diff
@@ -0,0 +1,197 @@
+	defer unlockRepo(srcLock)
+	if err != nil {
+		return err
+	}
+
+	dstLock, err := lockRepo(dstRepo)
+	defer unlockRepo(dstLock)
+	if err != nil {
+		return err
+	}
+
+	debug.Log("Loading source index")
+	if err := srcRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	debug.Log("Loading destination index")
+	if err := dstRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	for sn := range FindFilteredSnapshots(ctx, srcRepo, opts.Hosts, opts.Tags, opts.Paths, args) {
+		Verbosef("snapshot %s of %v at %s)\n", sn.ID().Str(), sn.Paths, sn.Time)
+		Verbosef("  copy started, this may take a while...\n")
+
+		if err := copySnapshot(ctx, srcRepo, dstRepo, *sn.Tree); err != nil {
+			return err
+		}
+		debug.Log("tree copied")
+
+		if err = dstRepo.Flush(ctx); err != nil {
+			return err
+		}
+		debug.Log("flushed packs")
+
+		err = dstRepo.SaveIndex(ctx)
+		if err != nil {
+			debug.Log("error saving index: %v", err)
+			return err
+		}
+		debug.Log("saved index")
+
+		// save snapshot
+		sn.Parent = nil   // Parent does not have relevance in the new repo.
+		sn.Original = nil // Original does not have relevance in the new repo.
+		newID, err := dstRepo.SaveJSONUnpacked(ctx, restic.SnapshotFile, sn)
+		if err != nil {
+			return err
+		}
+		Verbosef("snapshot %s saved\n", newID.Str())
+	}
+	return nil
+}
+
+func copySnapshot(ctx context.Context, srcRepo, dstRepo restic.Repository, treeID restic.ID) error {
+	tree, err := srcRepo.LoadTree(ctx, treeID)
+	if err != nil {
+		return fmt.Errorf("LoadTree(%v) returned error %v", treeID.Str(), err)
+	}
+	// Do we already have this tree blob?
+	if !dstRepo.Index().Has(treeID, restic.TreeBlob) {
+		newTreeID, err := dstRepo.SaveTree(ctx, tree)
+		if err != nil {
+			return fmt.Errorf("SaveTree(%v) returned error %v", treeID.Str(), err)
+		}
+		// Assurance only.
+		if newTreeID != treeID {
+			return fmt.Errorf("SaveTree(%v) returned unexpected id %s", treeID.Str(), newTreeID.Str())
+		}
+	}
+
+	// TODO: keep only one (big) buffer around.
+	// TODO: parellize this stuff, likely only needed inside a tree.
+
+	for _, entry := range tree.Nodes {
+		// If it is a directory, recurse
+		if entry.Type == "dir" && entry.Subtree != nil {
+			if err := copySnapshot(ctx, srcRepo, dstRepo, *entry.Subtree); err != nil {
+				return err
+			}
+		}
+
+		// Copy the blobs for this file.
+		var buf []byte
+		for _, blobID := range entry.Content {
+			// Do we already have this data blob?
+			if dstRepo.Index().Has(blobID, restic.DataBlob) {
+				continue
+			}
+			debug.Log("Copying blob %s\n", blobID.Str())
+			buf, err = srcRepo.LoadBlob(ctx, restic.DataBlob, blobID, buf)
+			if err != nil {
+				return fmt.Errorf("LoadBlob(%v) returned error %v", blobID, err)
+			}
+
+			newBlobID, err := dstRepo.SaveBlob(ctx, restic.DataBlob, buf, blobID)
+			if err != nil {
+				return fmt.Errorf("SaveBlob(%v) returned error %v", blobID, err)
+			}
+			// Assurance only.
+			if newBlobID != blobID {
```

Comment:
`SaveBlob` does not recalculate the blob hash, thus `blobID` and `newBlobID` will always be equal.